job-runner/chain/traverser.go

```diff
@@ -377,3 +386,36 @@ func (t *traverser) sendJL(pJob proto.Job, err error) {
 		jLogger.Errorf("problem sending job log (%#v) to the RM: %s", jl, err)
 	}
 }
+
+func (t *traverser) retrySequence(j proto.Job) {
+	sequenceStartJob := t.chain.SequenceStartJob(j)
+
+	// Create list of jobs in sequence that have been ran and we need to undo
+	var jobsToUndo []string
+	jobsToUndo = append(jobsToUndo, j.Id)
+	currSeqJobId := sequenceStartJob.Id
+	for currSeqJobId != j.Id {
+		jobsToUndo = append(jobsToUndo, currSeqJobId)
+		nextJobs := t.chain.NextJobs(currSeqJobId)
+		if len(nextJobs) != 1 {
+			// error
+		}
+		currSeqJobId = nextJobs[0].Id
+	}
+
+	// Set jobs state to pending so when sequence is retried they are ran again
+	for _, jobId := range jobsToUndo {
+		t.chain.SetJobState(jobId, proto.STATE_PENDING)
+	}
+
+	// Decrement count of jobs from total count of jobs done in chain
+	t.chain.n -= uint(len(jobsToUndo))
+
+	// Increment retry count for this sequence
+	t.chain.IncrementSequenceRetryCount(sequenceStartJob.Id)
+
+	t.sendJL(j, nil)
+
+	// Enqueue first job in sequence
+	t.runJobChan <- sequenceStartJob
```

Comment:
This could cause a panic. To see it, consider the implicit concurrency usage of this chan, and which chunk of code (a function) is responsible for the life and management, so to speak, of this chan.